### ISSUE TITLE
Unsigned negation in for inferred expressions doesn't warn

### ISSUE THREAD

**Issue by Aatch** [rust-lang/rust#24676]
While this warns about unsigned negation:

```rust
let x : u8 = -1;
```

This does not:

```rust
let x = -1;
let y : u8 = x;
```

This especially egregious in places like range expressions where it can cause a loop to not run at all when the lower bound is negative (and therefore becomes a large unsigned integer).
